src/main/java/com/hivemq/extensions/dns/metrics/DnsDiscoveryMetrics.java

```diff
@@ -0,0 +1,43 @@
+package com.hivemq.extensions.dns.metrics;
+
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.MetricRegistry;
+import com.hivemq.extension.sdk.api.annotations.NotNull;
+
+
+/**
+ * @author Lukas Brand
+ */
+public class DnsDiscoveryMetrics {
+
+    private final static @NotNull String DNS_DISCOVERY_EXTENSION = "dns-cluster-discovery-extension";
+    private final static @NotNull String HIVEMQ_PREFIX = "com.hivemq";
+
+    private final @NotNull MetricRegistry metricRegistry;
+    private final @NotNull Counter connectCounter;
+    private final @NotNull Counter failedConnectCounter;
+
+    public DnsDiscoveryMetrics(final @NotNull MetricRegistry metricRegistry) {
+        this.metricRegistry = metricRegistry;
+        this.connectCounter = metricRegistry.counter(MetricRegistry.name(HIVEMQ_PREFIX, DNS_DISCOVERY_EXTENSION, "resolution-request-success.count"));
+        this.failedConnectCounter = metricRegistry.counter(MetricRegistry.name(HIVEMQ_PREFIX, DNS_DISCOVERY_EXTENSION, "resolution-request-failed.count"));
+    }
+
+    public @NotNull Counter getResolutionRequestCounter() {
+        return connectCounter;
+    }
+
+    public @NotNull Counter getResolutionRequestFailedCounter() {
+        return failedConnectCounter;
+    }
+
+
+    public void registerAddressCountGauge(final @NotNull Gauge<Integer> supplier) {
```

Comment:
why not simply return a gauge here that can be held as a field and set in the discovery class? testability?

Reply:
The gauge itself is a functional interface which is set to an atomicInteger::get. In order to implement your approach I would need to return a MetricRegistry.MetricSupplier<com.codahale.metrics.Gauge> supplier which felt a bit weird and unreadable. Also, according to the documentation, calling the registerAddressCountGauge itself multiple times should not result in exceptions or unexpected behaviour. One could simply change the source of the Gauge. In case of calling the function with the same supplier it should not change the gauge at all.